### ISSUE TITLE
Weblog Post Metadata: Switch to Front Matter

### ISSUE THREAD

**Issue by RickStrahl** [RickStrahl/MarkdownMonster#175]
During recent work in the Weblog post interface with @thomaslevesque we've run into a few issues related  to properly persisting post metadata as it relates to posts.

It seems that Markdown in general tends to use YAML Front Matter style headers. 

When I originally built the post meta data YAML wasn't on my radar, and I opted to use pseudo XML for persistance so that it would be easy to edit text manually if necessary. The format used is not strict XML and is parsed as a string, so it's tolerant of invalid XML element formatting (ie. no problems with ", &, <, > etc. This is nice but as it turns out there are still ways to cause parsing problems.

Front Matter seems to be much more popular and in use with other text/documentation platforms.

So - the question for this discussion is whether we should switch to Front Matter meta data and if so what format and where should the Front Matter go. By convention Front Matter - as the name implies - goes in the front of the post. But this is somewhat noisy in that you have to sift through the front matter text before reading into the markdown.

Just for a heads up in regards to what metadata looks like, here's what a typical Front Matter header for a Web log post would look like:

```markdown
---
postId: 9022271
title: 'Tip: Create a Visual Studio Menu option to Open a Command Window'
abstract: 'This is the abstract for this blog post... '
keywords: Menu,External Tools, Command Line
categories: ASP.NET, .NET, Visual Studio
weblogName: West Wind Weblog
customFields:
  mt_markdown:
    iD: 1
    key: mt_markdown
    value: markdown
  wp_featured_image:
    iD: 1
    key: wp_featured_image
    value: http://blah.com/image.png
---

# Weblog Post starts here
asdaksd askldj alskdj aslkdj aslkdjasd

```

Trying to get a feel for whether there is interest one way or another to move this forward? 

If we do go ahead with this, should we keep the meta data at the top in typical YAML Front Matter style, or should we go the same route we did with the XML metadata that gets appended at the bottom of the document as a hidden comment that gets stripped out?

Any feedback would be welcome.

**Comment by thomaslevesque** [comment]
@RickStrahl I'm all for it, if you don't mind losing the metadata for existing documents (of course, it would also be possible to add code to migrate the XML metadata to Front Matter)

As for putting it at the top or bottom, I'd rather put it at the top, because:
1. It's standard
2. It's not much of a hassle to have to skip through it; it's pretty obvious when you look at it that it's not part of the actual document.
3. There's no safe way to locate it if it's not at the top; `---` already means something in Markdown, so you can't tell if it's intended to be a Front Matter block or normal Markdown.

**Comment by RickStrahl** [comment]
Yeah I feel the same.  I've hooked up the Yaml parsing logic but it's not plugged in at this point. We can most definitely use the old format as well at least for a while and if found just move it over to YAML.

3 - If it doesn't go on top it would go into a comment block the same way the XML is currently and it's then stripped out. But putting it at the top is probably better because most parsers understand the YAML header already and don't render it.

I'm working this in slowly, but not ready to push into the working code base just yet - it might be good to normalize the data to match standard Front Matter formatting which handle lists like tags a little differently.

**Comment by RickStrahl** [comment]
Implemented as of MM 1.2.17.
